File: ligature-test-suite/shared/src/main/scala/dev/ligature/testsuite/LigatureTestSuite.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.testsuite

import dev.ligature._
import munit.FunSuite

abstract class LigatureTestSuite extends FunSuite {
  def createLigature(): Ligature

  val testGraph = Graph("test/test")
  val testGraph2 = Graph("test/test2")
  val testGraph3 = Graph("test3/test")
  val a = Label("a")
  val b = Label("b")
  val label1 = Label("a")
  val label2 = Label("b")
  val label3 = Label("c")

  val setup = FunFixture[Ligature](
    setup = { test =>
      //Files.createTempFile("tmp", test.name)
      createLigature()
    },
    teardown = { instance =>
      instance.close()
    }
  )

  setup.test("create and close store".only) { instance =>
    assertEquals(instance.allGraphs().toList, List())
  }

  setup.test("creating a new graph".only) { instance =>
    instance.createGraph(testGraph)
    val res = instance.allGraphs().toList
    assertEquals(res, List(testGraph))
  }

  setup.test("check if graphs exist") { instance =>
    instance.createGraph(testGraph)
    val exists1 = instance.graphExists(testGraph)
    val exists2 = instance.graphExists(testGraph2)
    val res = (exists1, exists2)
    assertEquals(res, (true, false))
  }

  setup.test("match graphs prefix exact") { instance =>
    instance.createGraph(testGraph)
    val res = instance.matchGraphsPrefix("test/test").toList
    assertEquals(res.length, 1)
  }

  setup.test("match graphs prefix") { instance =>
    instance.createGraph(testGraph)
    instance.createGraph(testGraph2)
    instance.createGraph(testGraph3)
    val res1 = instance.matchGraphsPrefix("test").length
    val res2 = instance.matchGraphsPrefix("test/").length
    val res3 = instance.matchGraphsPrefix("snoo").length
    assertEquals((res1, res2, res3), (3, 2, 0))
  }

  setup.test("match graphs range") { instance =>
    instance.createGraph(Graph("a"))
    instance.createGraph(Graph("app"))
    instance.createGraph(Graph("b"))
    instance.createGraph(Graph("be"))
    instance.createGraph(Graph("bee"))
    instance.createGraph(
      Graph("test1/test")
    )
    instance.createGraph(
      Graph("test2/test2")
    )
    instance.createGraph(
      Graph("test3/test")
    )
    instance.createGraph(Graph("test4"))
    instance.createGraph(Graph("z"))
    instance.allGraphs().toList.length
    val res1 = instance.matchGraphsRange("a", "b").toList.length
    val res2 = instance.matchGraphsRange("be", "test3").toList.length
    val res3 = instance.matchGraphsRange("snoo", "zz").toList.length
    assertEquals((res1, res2, res3), (2, 4, 5)) // TODO check instances not just counts
  }

  setup.test("create and delete new graph") { instance =>
    instance.createGraph(testGraph)
    instance.deleteGraph(testGraph)
    instance.deleteGraph(testGraph2)
    val res = instance.allGraphs().toList
    assertEquals(res, List())
  }

  setup.test("new graphs should be empty") { instance =>
    instance.createGraph(testGraph)
    val res = instance.allEdges(testGraph).toList
    assertEquals(res, List())
  }

  setup.test("adding statements to graphs") { instance =>
      instance.createGraph(testGraph)
      instance.addEdges(
          testGraph,
            Seq(
              Edge(label1, a, label2),
              Edge(label1, a, label2), // dupe
              Edge(label1, a, label3)
            ).iterator
        )
        val edges = instance.allEdges(testGraph).toSet
      assertEquals(
        edges,
        Set(
          Edge(label1, a, label2),
          Edge(label1, a, label3)
        )
      )
    }

  setup.test("add Edge with IntegerLiteral Value") { instance =>
    instance.createGraph(testGraph)
    instance.addEdges(
      testGraph,
      Seq(
        Edge(label1, a, label2),
        Edge(label1, a, LigatureLiteral.IntegerLiteral(100)),
        Edge(label1, a, LigatureLiteral.IntegerLiteral(101)),
        Edge(label1, a, LigatureLiteral.IntegerLiteral(100)),
        Edge(label2, a, LigatureLiteral.IntegerLiteral(-243729))
      ).iterator
    )
    val edges = instance.allEdges(testGraph).toSet
    assertEquals(
      edges,
      Set(
        Edge(label1, a, label2),
        Edge(label1, a, LigatureLiteral.IntegerLiteral(100)),
        Edge(label1, a, LigatureLiteral.IntegerLiteral(101)),
        Edge(label2, a, LigatureLiteral.IntegerLiteral(-243729))
      )
    )
  }

  setup.test("add Edge with StringLiteral Value") { instance =>
    instance.createGraph(testGraph)
    instance.addEdges(
      testGraph,
        Seq(
          Edge(label1, a, label2),
          Edge(label1, a, LigatureLiteral.StringLiteral("text")),
          Edge(label1, a, LigatureLiteral.StringLiteral("text2")),
          Edge(label1, a, LigatureLiteral.StringLiteral("text")),
          Edge(label2, a, LigatureLiteral.StringLiteral("text"))
        ).iterator
    )
    val edges = instance.allEdges(testGraph).toSet
    assertEquals(
      edges,
      Set(
        Edge(label1, a, label2),
        Edge(label1, a, LigatureLiteral.StringLiteral("text")),
        Edge(label1, a, LigatureLiteral.StringLiteral("text2")),
        Edge(label2, a, LigatureLiteral.StringLiteral("text"))
      )
    )
  }

// //   // setup.test("new identifiers") {
// //   //   val instance = createLigature
// //   //   val res = for {
// //   //     _ <- instance.createGraph(testGraph)
// //   //     _ <- instance.write(testGraph) { tx =>
// //   //       for {
// //   //         entity <- tx.newLabel("entity-")
// //   //         attribute <- tx.newLabel("attribute-")
// //   //         value <- tx.newLabel("value-")
// //   //         _ <- tx.addEdge(Edge(entity, attribute, value))
// //   //       } yield IO.unit
// //   //     }
// //   //     statements <- instance.query(testGraph) { tx =>
// //   //       tx.allEdges().toList
// //   //     }
// //   //   } yield statements.head
// //   //   res.map { it =>
// //   //     assert(it.entity.name.startsWith("entity-"))
// //   //     assert(it.attribute.name.startsWith("attribute-"))
// //   //     it.value match {
// //   //       case Label(id) => assert(id.startsWith("value-"))
// //   //       case _              => assert(false)
// //   //     }
// //   //   }
// //   // }

  setup.test("removing statements from graphs") { instance =>
    instance.createGraph(testGraph)
    instance.addEdges(testGraph, Seq(Edge(label1, a, label2)).iterator)
    instance.removeEdges(
      testGraph,
      Seq(Edge(label1, a, label2), Edge(label1, a, label2)).iterator
    )
    val edges = instance.allEdges(testGraph).toSet
    assertEquals(edges, Set())
  }

  setup.test("removing statements from graphs with dupe") { instance =>
    instance.createGraph(testGraph)
    instance.addEdges(
      testGraph,
      Seq(Edge(label1, a, label2), Edge(label3, a, label2), Edge(label1, a, label2)).iterator
    )
    instance.removeEdges(testGraph, Seq(Edge(label1, a, label2)).iterator)
    val edges = instance.allEdges(testGraph).toSet
    assertEquals(edges, Set(Edge(label3, a, label2)))
  }

  setup.test("removing statements from graphs with duplicate Strings") { instance =>
    instance.createGraph(testGraph)
    instance.addEdges(
      testGraph,
        Seq(
          Edge(label1, a, LigatureLiteral.StringLiteral("hello")),
          Edge(label1, a, LigatureLiteral.StringLiteral("hello")),
          Edge(label2, a, LigatureLiteral.StringLiteral("hello"))
        ).iterator
    )
    instance.removeEdges(
      testGraph,
      Seq(Edge(label1, a, LigatureLiteral.StringLiteral("hello"))).iterator
    )
    val edges = instance.allEdges(testGraph).toSet
    assertEquals(edges, Set(Edge(label2, a, LigatureLiteral.StringLiteral("hello"))))
  }

// //   // setup.test("allow canceling WriteTx by throwing exception") {
// //   //     val res = createLigature.instance.use { instance  =>
// //   //         for {
// //   //             _ <- instance.createGraph(testGraph)
// //   //             _ <- instance.write(testGraph).use { tx =>
// //   //                 for {
// //   //                     _ <- tx.addEdge(Edge(label1, a, label2))
// //   //                     _ <- tx.cancel()
// //   //                 } yield ()
// //   //             }
// //   //             _ <- instance.write(testGraph).use { tx =>
// //   //                 for {
// //   //                     _ <- tx.addEdge(Edge(label2, a, label3))
// //   //                     _ <- tx.addEdge(Edge(label3, a, label2))
// //   //                 } yield ()
// //   //             }
// //   //             statements <- instance.query(testGraph).use { tx =>
// //   //                 tx.allEdges().toList
// //   //             }
// //   //         } yield statements
// //   //     }.unsafeRunSync().map(_.right.get).map(_.statement).toSet
// //   //     assertEquals(res, Set(
// //   //         Edge(label2, a, label3),
// //   //         Edge(label3, a, label2)))
// //   // }

  setup.test("matching statements in graphs") { instance =>
    instance.createGraph(testGraph)
    instance.addEdges(
      testGraph,
        Seq(
          Edge(label1, a, LigatureLiteral.StringLiteral("Hello")),
          Edge(label2, a, label1),
          Edge(label2, a, label3),
          Edge(label3, b, label2),
          Edge(label3, b, LigatureLiteral.StringLiteral("Hello"))
        ).iterator
    )
    val (all, as, hellos, helloa) = instance.query(testGraph) { tx =>
      val all = tx.matchEdges().toSet
      val as = tx.matchEdges(None, Some(a)).toSet
      val hellos = tx
          .matchEdges(None, None, Some(LigatureLiteral.StringLiteral("Hello")))
          .toSet
      val helloa = tx
          .matchEdges(None, Some(a), Some(LigatureLiteral.StringLiteral("Hello")))
          .toSet
      (all, as, hellos, helloa)
    }
    assertEquals(
      all,
      Set(
        Edge(label1, a, LigatureLiteral.StringLiteral("Hello")),
        Edge(label2, a, label1),
        Edge(label2, a, label3),
        Edge(label3, b, label2),
        Edge(label3, b, LigatureLiteral.StringLiteral("Hello"))
      )
    )
    assertEquals(
      as,
      Set(
        Edge(label1, a, LigatureLiteral.StringLiteral("Hello")),
        Edge(label2, a, label1),
        Edge(label2, a, label3)
      )
    )
    assertEquals(
      hellos,
      Set(
        Edge(label1, a, LigatureLiteral.StringLiteral("Hello")),
        Edge(label3, b, LigatureLiteral.StringLiteral("Hello"))
      )
    )
    assertEquals(
      helloa,
      Set(
        Edge(label1, a, LigatureLiteral.StringLiteral("Hello"))
      )
    )
  }

// //  setup.test("matching statements with literals and ranges in graphs") {
// //    val instance = createLigature
// //    val res = for {
// //      _ <- instance.createGraph(testGraph)
// //      _ <- instance.write(testGraph) { tx =>
// //        for {
// //          _ <- tx.addEdge(Edge(label1, a, label2))
// //          _ <- tx
// //            .addEdge(Edge(label1, b, StringLiteral("add")))
// //          _ <- tx
// //            .addEdge(Edge(label1, a, IntegerLiteral(5L)))
// //          _ <- tx
// //            .addEdge(Edge(label2, a, IntegerLiteral(3L)))
// //          _ <- tx.addEdge(
// //            Edge(label2, a, StringLiteral("divide"))
// //          )
// //          _ <- tx.addEdge(Edge(label2, b, label3))
// //          _ <- tx
// //            .addEdge(Edge(label3, a, IntegerLiteral(7L)))
// //          _ <- tx.addEdge(
// //            Edge(label3, b, StringLiteral("decimal"))
// //          )
// //        } yield ()
// //      }
// //      res <- instance.query(testGraph) { tx =>
// //        for {
// //          res1 <- tx
// //            .matchEdgesRange(None, None, StringLiteralRange("a", "dd"))
// //            .compile
// //            .toList
// //          res2 <- tx
// //            .matchEdgesRange(None, None, IntegerLiteralRange(3, 6))
// //            .compile
// //            .toList
// //          res3 <- tx
// //            .matchEdgesRange(None, Some(b), StringLiteralRange("ae", "df"))
// //            .compile
// //            .toList
// //        } yield (res1.toSet, res2.toSet, res3.toSet)
// //      }
// //    } yield res
// //    res.map { (res1, res2, res3) =>
// //      assertEquals(
// //        res1,
// //        Set(
// //          Edge(label1, b, StringLiteral("add"))
// //        )
// //      )
// //      assertEquals(
// //        res2,
// //        Set(
// //          Edge(label2, a, IntegerLiteral(3L)),
// //          Edge(label1, a, IntegerLiteral(5L))
// //        )
// //      )
// //      assertEquals(
// //        res3,
// //        Set(
// //          Edge(label3, b, StringLiteral("decimal"))
// //        )
// //      )
// //    }
// //  }
}

File: ligature-test-suite/src/commonMain/kotlin/dev/ligature/testsuite/LigatureTestSuite.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.testsuite

import dev.ligature.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldStartWith
import arrow.core.none
import arrow.core.Some

abstract class LigatureTestSuite: FunSpec() {
  abstract fun createLigature(): Ligature

  val testDataset = Dataset("test/test")
  val testDataset2 = Dataset("test/test2")
  val testDataset3 = Dataset("test3/test")
  val a = Identifier("a")
  val b = Identifier("b")
  val entity1 = Identifier("a")
  val entity2 = Identifier("b")
  val entity3 = Identifier("c")

  fun runTest(fn: suspend (Ligature) -> Unit) {
    val ligature = createLigature()
    runBlocking {
      fn(ligature)
      ligature.close()
    }
  }

  init {
    test("create and close store") {
      runTest { ligature ->
        ligature.allDatasets().toList() shouldBe listOf()
      }
    }

    test("creating a new dataset") {
      runTest { ligature ->
        ligature.createDataset(testDataset)
        ligature.allDatasets().toList() shouldBe listOf(testDataset)
      }
    }

    test("check if datasets exist") {
      runTest { ligature ->
        ligature.createDataset(testDataset)
        ligature.datasetExists(testDataset) shouldBe true
        ligature.datasetExists(testDataset2) shouldBe false
      }
    }

    test("match datasets prefix exact") {
      runTest {  ligature ->
        ligature.createDataset(testDataset)
        ligature.matchDatasetsPrefix("test/test").toList().size shouldBe 1
      }
    }

    test("match datasets prefix") {
      runTest { ligature ->
        ligature.createDataset(testDataset)
        ligature.createDataset(testDataset2)
        ligature.createDataset(testDataset3)
        ligature.matchDatasetsPrefix("test").toList().size shouldBe 3
        ligature.matchDatasetsPrefix("test/").toList().size shouldBe 2
        ligature.matchDatasetsPrefix("snoo").toList().size shouldBe 0
      }
    }

  test("match datasets range") {
    runTest { ligature ->
      ligature.createDataset(Dataset("a"))
      ligature.createDataset(Dataset("app"))
      ligature.createDataset(Dataset("b"))
      ligature.createDataset(Dataset("be"))
      ligature.createDataset(Dataset("bee"))
      ligature.createDataset(Dataset("test1/test"))
      ligature.createDataset(Dataset("test2/test2"))
      ligature.createDataset(Dataset("test3/test"))
      ligature.createDataset(Dataset("test4"))
      ligature.createDataset(Dataset("z"))
      ligature.matchDatasetsRange("a", "b").toList().size shouldBe 2
      ligature.matchDatasetsRange("be", "test3").toList().size shouldBe 4
      ligature.matchDatasetsRange("snoo", "zz").toList().size shouldBe 5
    }
  }

  test("create and delete new dataset") {
    runTest { ligature ->
      ligature.createDataset(testDataset)
      ligature.deleteDataset(testDataset)
      ligature.deleteDataset(testDataset2)
      ligature.allDatasets().toList() shouldBe listOf()
    }
  }

  test("new datasets should be empty") {
    runTest { ligature ->
      ligature.createDataset(testDataset)
      ligature.query(testDataset) { tx -> tx.allStatements().toList() } shouldBe listOf()
    }
  }

  test("adding statements to datasets") {
    runTest { ligature ->
      ligature.createDataset(testDataset)
      ligature.write(testDataset) { tx ->
        tx.addStatement(Statement(entity1, a, entity2))
        tx.addStatement(Statement(entity1, a, entity2)) // dupe
        tx.addStatement(Statement(entity1, a, entity3))
      }
      ligature.query(testDataset) { tx ->
        tx.allStatements().toSet()
      } shouldBe setOf(
        Statement(entity1, a, entity2),
        Statement(entity1, a, entity3)
      )
    }
  }

  test("add Statement with IntegerLiteral Value") {
    runTest {  ligature ->
        ligature.createDataset(testDataset)
        ligature.write(testDataset) { tx ->
          tx.addStatement(Statement(entity1, a, entity2))
          tx.addStatement(Statement(entity1, a, IntegerLiteral(100)))
          tx.addStatement(Statement(entity1, a, IntegerLiteral(101)))
          tx.addStatement(Statement(entity1, a, IntegerLiteral(100)))
          tx.addStatement(Statement(entity2, a, IntegerLiteral(-243729)))
        }
        ligature.query(testDataset) { tx ->
          tx.allStatements().toSet()
        } shouldBe setOf(
          Statement(entity1, a, entity2),
          Statement(entity1, a, IntegerLiteral(100)),
          Statement(entity1, a, IntegerLiteral(101)),
          Statement(entity2, a, IntegerLiteral(-243729))
        )
    }
  }

  test("add Statement with StringLiteral Value") {
    runTest { ligature ->
      ligature.createDataset(testDataset)
      ligature.write(testDataset) { tx ->
        tx.addStatement(Statement(entity1, a, entity2))
        tx.addStatement(Statement(entity1, a, StringLiteral("text")))
        tx.addStatement(Statement(entity1, a, StringLiteral("text2")))
        tx.addStatement(Statement(entity1, a, StringLiteral("text")))
        tx.addStatement(Statement(entity2, a, StringLiteral("text")))
      }
      ligature.query(testDataset) { tx ->
        tx.allStatements().toSet()
      } shouldBe setOf(
        Statement(entity1, a, entity2),
        Statement(entity1, a, StringLiteral("text")),
        Statement(entity1, a, StringLiteral("text2")),
        Statement(entity2, a, StringLiteral("text"))
      )
    }
  }

  test("new identifiers") {
    runTest { ligature ->
      ligature.createDataset(testDataset)
      ligature.write(testDataset) { tx ->
        val entity = tx.newIdentifier("entity-")
        val attribute = tx.newIdentifier("attribute-")
        val value = tx.newIdentifier("value-")
        tx.addStatement(Statement(entity, attribute, value))
      }
      val statement = ligature.query(testDataset) { tx ->
        tx.allStatements().toList().first()
      }
      statement.entity.name.shouldStartWith("entity-")
      statement.attribute.name.shouldStartWith("attribute-")
      when(val value = statement.value) {
        is Identifier -> value.name.shouldStartWith("value-")
        else          -> throw Exception("Failed test.")
      }
    }
  }

  test("removing statements from datasets") {
    runTest { ligature ->
      ligature.createDataset(testDataset)
      ligature.write(testDataset) { tx ->
        tx.addStatement(Statement(entity1, a, entity2))
        tx.removeStatement(Statement(entity1, a, entity2))
        tx.removeStatement(Statement(entity1, a, entity2))
      }
      ligature.query(testDataset) { tx ->
        tx.allStatements().toList() shouldBe listOf()
      }
    }
  }

  test("removing statements from datasets with dupe") {
    runTest { ligature ->
      ligature.createDataset(testDataset)
      ligature.write(testDataset) { tx ->
        tx.addStatement(Statement(entity1, a, entity2))
        tx.addStatement(Statement(entity3, a, entity2))
        tx.removeStatement(Statement(entity1, a, entity2))
        tx.removeStatement(Statement(entity1, a, entity2))
      }
      ligature.query(testDataset) { tx ->
        tx.allStatements().toList()
      } shouldBe listOf(Statement(entity3, a, entity2))
    }
  }

  test("removing statements from datasets with duplicate Strings") {
    runTest { ligature ->
      ligature.createDataset(testDataset)
      ligature.write(testDataset) { tx ->
        tx.addStatement(Statement(entity1, a, StringLiteral("hello")))
        tx.addStatement(Statement(entity1, a, StringLiteral("hello")))
        tx.addStatement(Statement(entity2, a, StringLiteral("hello")))
        tx.removeStatement(Statement(entity1, a, StringLiteral("hello")))
      }
      ligature.query(testDataset) { tx ->
        tx.allStatements().toList()
      } shouldBe listOf(Statement(entity2, a, StringLiteral("hello")))
    }
  }

//  // test("allow canceling WriteTx by throwing exception") {
//  //     val res = createLigature.instance.use { instance  ->
//  //         for {
//  //             ligature.createDataset(testDataset)
//  //             ligature.write(testDataset).use { tx ->
//  //                 for {
//  //                     tx.addStatement(Statement(entity1, a, entity2))
//  //                     tx.cancel()
//  //                 } yield ()
//  //             }
//  //             ligature.write(testDataset).use { tx ->
//  //                 for {
//  //                     tx.addStatement(Statement(entity2, a, entity3))
//  //                     tx.addStatement(Statement(entity3, a, entity2))
//  //                 } yield ()
//  //             }
//  //             statements <- instance.query(testDataset).use { tx ->
//  //                 tx.allStatements().toList()
//  //             }
//  //         } yield statements
//  //     }.unsafeRunSync().map(_.right.get).map(_.statement).toSet
//  //     assertEquals(res, Set(
//  //         Statement(entity2, a, entity3),
//  //         Statement(entity3, a, entity2)))
//  // }

  test("matching statements in datasets") {
    runTest { ligature ->
      ligature.createDataset(testDataset)
      ligature.write(testDataset) { tx ->
        tx.addStatement(
          Statement(entity1, a, StringLiteral("Hello"))
        )
        tx.addStatement(Statement(entity2, a, entity1))
        tx.addStatement(Statement(entity2, a, entity3))
        tx.addStatement(Statement(entity3, b, entity2))
        tx.addStatement(
          Statement(entity3, b, StringLiteral("Hello"))
        )
      }
      ligature.query(testDataset) { tx ->
          tx.matchStatements().toSet() shouldBe setOf(
            Statement(entity1, a, StringLiteral("Hello")),
            Statement(entity2, a, entity1),
            Statement(entity2, a, entity3),
            Statement(entity3, b, entity2),
            Statement(entity3, b, StringLiteral("Hello"))
          )

          tx.matchStatements(null, a).toSet() shouldBe setOf(
            Statement(entity1, a, StringLiteral("Hello")),
            Statement(entity2, a, entity1),
            Statement(entity2, a, entity3)
          )

          tx.matchStatements(null, null, StringLiteral("Hello"))
              .toSet() shouldBe setOf(
                Statement(entity1, a, StringLiteral("Hello")),
                Statement(entity3, b, StringLiteral("Hello"))
              )

          tx.matchStatements(null, a, StringLiteral("Hello"))
            .toSet() shouldBe setOf(
              Statement(entity1, a, StringLiteral("Hello"))
            )
        }
      }
    }
////  test("matching statements with literals and ranges in datasets") {
////    val instance = createLigature
////    val res = for {
////      ligature.createDataset(testDataset)
////      ligature.write(testDataset) { tx ->
////        for {
////          tx.addStatement(Statement(entity1, a, entity2))
////          tx
////            .addStatement(Statement(entity1, b, StringLiteral("add")))
////          tx
////            .addStatement(Statement(entity1, a, IntegerLiteral(5L)))
////          tx
////            .addStatement(Statement(entity2, a, IntegerLiteral(3L)))
////          tx.addStatement(
////            Statement(entity2, a, StringLiteral("divide"))
////          )
////          tx.addStatement(Statement(entity2, b, entity3))
////          tx
////            .addStatement(Statement(entity3, a, IntegerLiteral(7L)))
////          tx.addStatement(
////            Statement(entity3, b, StringLiteral("decimal"))
////          )
////        } yield ()
////      }
////      res <- instance.query(testDataset) { tx ->
////        for {
////          res1 <- tx
////            .matchStatementsRange(none(), none(), StringLiteralRange("a", "dd"))
////
////            .toList()
////          res2 <- tx
////            .matchStatementsRange(none(), none(), IntegerLiteralRange(3, 6))
////
////            .toList()
////          res3 <- tx
////            .matchStatementsRange(none(), Some(b), StringLiteralRange("ae", "df"))
////
////            .toList()
////        } yield (res1.toSet, res2.toSet, res3.toSet)
////      }
////    } yield res
////    res.map { (res1, res2, res3) ->
////      assertEquals(
////        res1,
////        Set(
////          Statement(entity1, b, StringLiteral("add"))
////        )
////      )
////      assertEquals(
////        res2,
////        Set(
////          Statement(entity2, a, IntegerLiteral(3L)),
////          Statement(entity1, a, IntegerLiteral(5L))
////        )
////      )
////      assertEquals(
////        res3,
////        Set(
////          Statement(entity3, b, StringLiteral("decimal"))
////        )
////      )
////    }
////  }
  }
}
